text/0000-rustdoc-cargo-configuration.md

```diff
@@ -0,0 +1,412 @@
+    "config-file-arg": "--config-file",
+    "accepts-spanned-json": true
+},
+{   // sample clippy config.
+    "tool": "clippy",
+    "cargo-tables": ["lints"],
+    "search-files": ["clippy.toml", ".clippy.toml"]
+    // ...
+}
+// ... any tools just needs to provide a similar schema
+```
+
+The idea is that a build system could store this configuration for `rustdoc`,
+`rustfmt`, `clippy`, or some other tool, and know exactly how to invoke it. A
+package on `crates.io` could potentially include this information as part of its
+`Cargo.toml`, and Cargo would be able to invoke it with custom configuration.
+
+The schema is as follows:
+
+-   `tool`: name of the tool, which can be configured by a `[tools.toolname]`
+    table. Contents of `[tools.toolname]` will be passed as top-level JSON keys.
+-   `manifest-tables`: tables other than `[tool.toolname]` that should be passed
+-   `config-json-arg`: the CLI flag to pass JSON from Cargo or other build systems.
+    Must accept either a JSON string or a path to a JSON file.
+-   `config-files`: Files that Cargo should search for. Path is relative to a
+    `Cargo.toml` file; accepts relative paths and globs.
+-   `config-file-arg`: Argument to pass the paths of found files
+-   `accepts-spanned-json`: Whether or not Cargo should include spans when
+    serializing from a TOML file.
+
+A tool must follow the following rules:
+
+-   `config-json-arg` must accept either a JSON string or a file path
+-   `config-json-arg` and `config-file-arg` values must be able to be specified
+    more than once. A tool is responsible for merging these configurations
+-   When receiving multiple `config-json-arg` or `config-file-arg` arguments,
+    the first has lowest precedence and the last has the highest precedence.
+    That is, configuration specified by a later argument should overwrite
+    configuration from an earlier argument.
+
+Cargo, in turn, must follow the following process:
+
+1. Serialize `[tool.toolname]` and any requested `manifest-tables` from the
+   workspace `Cargo.toml`. Retain spans as specified by `accepts-spanned-json`.
+   Pass this information via the `config-json-arg` argument.
+2. Search at the workspace root for any paths matching `config-files`. Pass
+   these paths via `config-file-arg`. Order is not important.
+3. Repeat step 1 at crate root
+4. Repeat step 2 at crate root
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+-   This adds complexity to both Cargo and `rustdoc`: additional flags for
+    `rustdoc`, additional behavior for both.
+-   Obtaining information from `Cargo.toml`, such as feature descriptions, does
+    bring `rustdoc` and Cargo closer together when there have been advantages to
+    keeping the tools separate. This RFC attempts to mitigate this risk by not
+    locking the configuration to Cargo in any way, but it is still a concern.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
Could you add an alternative of using `package.metadata` / `workspace.metadata`?  What justifies adding a new table for this?

Reply:
I will update this, but my loose thought is that `package.metadata`/`workspace.metadata` is more for tools that read `Cargo.toml` themselves, whereas `tools.x` would be for tools that Cargo invokes. The other "soft" reason is that `package.metadata.rustdoc` seems like fairly deep configuration nesting for an official rust tool: `tools.clippy` or `tools.rustdoc` seem more fitting for something with first class support that `Cargo` is aware of.

(I will express the above in this RFC, but don't personally have a problem with `package.metadata` if the justification isn't sufficient).

Reply:
This does of course relate to `[lints]` and some of the decisions made there. I believe that as `rustdoc` configuration isn't a lint that putting this information in the `[lints]` table doesn't make sense... but it does very much run into the same questions as in https://github.com/rust-lang/rfcs/pull/3389#issuecomment-1431744220

Reply:
On relation to the lints RFC, the schema loosely proposed [in this section](https://github.com/tgross35/rfcs/blob/rustdoc-cargo-config/text/0000-rustdoc-cargo-configuration.md#extensible-configuration) could potentially also specify usage of the `[lints.x]` table for a generic tool. Hm...